plugins/audio-s5gen2/fu-audio-s5gen2-ble-device.c

```diff
@@ -0,0 +1,615 @@
+			       guint8 *data_in,
+			       gsize data_len,
+			       gsize *read_len,
+			       GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(impl);
+	gsize bufsz = 0;
+	g_autofree guint8 *buf = NULL;
+
+	bufsz = ((gsize)self->mtu < data_len + FU_QC_S5GEN2_GAIA_V3_HDR_SZ)
+		    ? (gsize)self->mtu
+		    : data_len + FU_QC_S5GEN2_GAIA_V3_HDR_SZ;
+
+	buf = g_malloc0(bufsz);
+
+	if (!fu_io_channel_read_raw(self->io_cmd,
+				    buf,
+				    bufsz,
+				    read_len,
+				    FU_QC_S5GEN2_BLE_DEVICE_TIMEOUT,
+				    FU_IO_CHANNEL_FLAG_SINGLE_SHOT,
+				    error))
+		return FALSE;
+
+	fu_dump_raw(G_LOG_DOMAIN, "Read from device:", buf, *read_len);
+	if (*read_len < FU_QC_S5GEN2_GAIA_V3_HDR_SZ) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "got %lu bytes, less than GAIA header",
+			    *read_len);
+		return FALSE;
+	}
+	memcpy(data_in,
+	       buf + FU_QC_S5GEN2_GAIA_V3_HDR_SZ,
+	       (*read_len) - FU_QC_S5GEN2_GAIA_V3_HDR_SZ);
+
+	return TRUE;
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_req_connect(FuQcS5gen2Impl *impl, GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(impl);
+	guint8 buf[FU_QC_S5GEN2_BLE_DEVICE_BUFFER_SZ] = {0};
+	gsize read_len;
+	g_autoptr(GByteArray) req = fu_struct_qc_gaia_v3_upgrade_connect_cmd_new();
+	g_autoptr(GByteArray) validate = NULL;
+
+	fu_struct_qc_gaia_v3_upgrade_connect_cmd_set_vendor_id(req, self->vid_v3);
+
+	if (!fu_qc_s5gen2_ble_device_notify_acquire(self, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_send(self, req->data, req->len, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_recv(self, buf, sizeof(buf), &read_len, error))
+		return FALSE;
+
+	validate = fu_struct_qc_gaia_v3_upgrade_connect_ack_parse(buf, read_len, 0, error);
+	if (validate == NULL)
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_qc_s5gen2_ble_device_req_disconnect(FuQcS5gen2Impl *impl, GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(impl);
+	guint8 buf[FU_QC_S5GEN2_BLE_DEVICE_BUFFER_SZ] = {0};
+	gsize read_len;
+	g_autoptr(GByteArray) req = fu_struct_qc_gaia_v3_upgrade_disconnect_cmd_new();
+	g_autoptr(GByteArray) validate = NULL;
+
+	fu_struct_qc_gaia_v3_upgrade_disconnect_cmd_set_vendor_id(req, self->vid_v3);
+
+	if (!fu_qc_s5gen2_ble_device_send(self, req->data, req->len, error))
+		return FALSE;
+
+	if (!fu_qc_s5gen2_ble_device_recv(self, buf, sizeof(buf), &read_len, error))
+		return FALSE;
+
+	validate = fu_struct_qc_gaia_v3_upgrade_disconnect_ack_parse(buf, read_len, 0, error);
+	if (validate == NULL)
+		return FALSE;
+
+	return fu_qc_s5gen2_ble_device_notify_release(self, error);
+}
+
+static gsize
+fu_qc_s5gen2_ble_device_data_size(FuQcS5gen2Impl *impl, GError **error)
+{
+	FuQcS5gen2BleDevice *self = FU_QC_S5GEN2_BLE_DEVICE(impl);
+	/* TODO: atm for GAIA only */
+	return self->mtu - FU_STRUCT_QC_DATA_SIZE - FU_QC_S5GEN2_GAIA_V3_HDR_SZ - 3;
```

Comment:
this has got "coverity screaming at me" written all over it. Can we check that self->mtu is bigger than (FU_STRUCT_QC_DATA_SIZE - FU_QC_S5GEN2_GAIA_V3_HDR_SZ - 3) first, and if it's smaller set a GError pls

Reply:
I'll set it to zero here, and check in fu_qc_s5gen2_device_write_bucket() there it is used.